File: tools/hod2lib/stage.py
"""What a *stage* is: the single abstraction every exporter resolves against.

A stage is not a directory and not a filename glob. It is a **scene id** --
the event system's own index -- and everything else follows from tables
compiled into ``Hod2.exe``:

    scene ---> evt/ file          0x00579928 -> 0x004D1C7C
          ---> route table        0x00597890      block flow graph
          ---> region tables      0x00576A2C/0x00576A5C   (mode 1: 0x00576A8C/ABC)
          ---> geometry set       union of every region's asset slots,
                                  plus every slot opcode 0x50 loads
          ---> cam/ files         cp_st<N>, op_st<N>

This module holds that resolution once. ``export_level.py`` and
``export_player.py`` both consume it, so the glTF a human opens in Blender and
the bundle the browser player eats are guaranteed to describe the same set of
models, regions and draw modes. Duplicating the resolution is exactly how the
two would drift.

Reference: docs/formats/pipeline.md.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path as _Path

from . import cam as camlib, coli as colilib, container as C, evt, exetab, nl1, texbank
from .campaths import CamPaths

__all__ = ["STAGE_TO_SCENE", "SCENE_TO_STAGE", "Stage", "Part",
           "get_tables", "load_asset", "load_cam_paths"]

#: stage number -> scene id, the event system's own index.
STAGE_TO_SCENE = {1: 0, 2: 1, 3: 2, 4: 3, 5: 4, 6: 5}
SCENE_TO_STAGE = {v: k for k, v in STAGE_TO_SCENE.items()}

#: Reading Hod2.exe's tables costs a second or two; a process rarely wants
#: more than one game directory, and every exporter wants the tables.
_TABLES: dict[str, exetab.ExeTables] = {}


def get_tables(game: _Path) -> exetab.ExeTables | None:
    """Parsed ``Hod2.exe`` tables for a game directory, cached per process."""
    key = str(game)
    if key not in _TABLES:
        exe = _Path(game) / "Hod2.exe"
        if not exe.exists():
            return None
        _TABLES[key] = exetab.ExeTables(exe)
    return _TABLES.get(key)


@dataclass
class Part:
    """One ``pol/`` file's contribution to a stage."""

    name: str                 #: file stem, e.g. ``st2_07``
    models: list = field(default_factory=list)
    bank: object | None = None

    def __iter__(self):
        # Exporters take parts as (name, models, bank) tuples.
        return iter((self.name, self.models, self.bank))


def _bank_for(game: _Path, stem: str):
    """The ``tex/`` bank paired with a ``pol/`` file, or None.

    The EXE's descriptor table is authoritative; the prefix-sum solver is only
    a fallback for banks it does not list.
    """
    tex = game / "tex" / f"{stem}.bin"
    if not tex.exists():
        return None
    tc = C.load(tex.read_bytes())
    data = tc.data if tc.kind == C.COMPRESSED else tex.read_bytes()
    tables = get_tables(game)
    entries = tables.entries(stem) if tables else []
    if entries:
        return texbank.bank_from_exe(data, entries)
    return None


def load_asset(game: _Path, name: str):
    """Return ``(models, bank)`` for a ``pol/`` asset and its ``tex/`` bank."""
    pol = game / "pol" / f"{name}.bin"
    if not pol.exists():
        raise FileNotFoundError(f"no such asset: {pol}")

    c = C.load(pol.read_bytes())
    if not c.models:
        raise ValueError(f"{name}: no models ({c.kind})")
    models = nl1.parse_container(c)

    bank = None
    tex = game / "tex" / f"{name}.bin"
    if tex.exists():
        tc = C.load(tex.read_bytes())
        data = tc.data if tc.kind == C.COMPRESSED else tex.read_bytes()
        tables = get_tables(game)
        entries = tables.entries(name) if tables else []
        if entries:
            bank = texbank.bank_from_exe(data, entries)
        else:
            descs = texbank.harvest_descriptors(models)
            if descs:
                bank = texbank.solve_layout(data, descs)
    return models, bank


def load_cam_paths(game: _Path, stage: int | None, name: str | None):
    """The ``cam/`` files belonging to a stage: ``cp_stN`` then ``op_stN``.

    Returns ``[]`` when the asset has no matching camera file, which is
    normal -- only the six stages and a handful of cutscenes have one.
    """
    stems: list[str] = []
    if stage is not None:
        stems = [f"cp_st{stage}", f"op_st{stage}"]
    elif name:
        m = re.match(r"st(\d+)_", name)      # st2_07 -> stage 2
        if m:
            stems = [f"cp_st{m.group(1)}", f"op_st{m.group(1)}"]

    out = []
    for stem in stems:
        p = _Path(game) / "cam" / f"{stem}.bin"
        if p.exists():
            out.append(camlib.load(str(p)))
    return out


class Stage:
    """One stage of the game, resolved from ``Hod2.exe``.

    Everything is lazy: constructing a ``Stage`` only reads the EXE tables, so
    it is cheap to ask what a stage *is* without paying to decode its
    geometry.
    """

    def __init__(self, game: _Path | str, *, stage: int | None = None,
                 scene: int | None = None, original: bool = False):
        self.game = _Path(game).expanduser().resolve()
        if scene is None:
            if stage is None:
                raise ValueError("give stage or scene")
            scene = STAGE_TO_SCENE.get(stage)
            if scene is None:
                raise ValueError(f"stage {stage} has no scene")
        self.scene = scene
        self.stage = stage if stage is not None else SCENE_TO_STAGE.get(scene)
        self.original = original
        self.game_mode = 1 if original else 0

        tables = get_tables(self.game)
        if tables is None:
            raise FileNotFoundError(
                f"Hod2.exe is required to resolve a stage: {self.game}")
        self.tables = tables

        self._geometry = None
        self._cam_files = None
        self._campaths = None
        self._evt = None
        self._colisets = None

    # -- identity ---------------------------------------------------------

    @property
    def name(self) -> str:
        """Bundle/export name: ``stage2``, or ``stage2_original``."""
        base = f"stage{self.stage}" if self.stage else f"scene{self.scene}"
        return base + ("_original" if self.original else "")

    @property
    def evt_file(self) -> str | None:
        return self.tables.scene_evt_file(self.scene)

    @property
    def routes(self) -> list[tuple[int, int, int, int]]:
        """``[(kind, next0, next1, next2), ...]`` -- the block flow graph."""
        return self.tables.scene_routes(self.scene)

    @property
    def block_count(self) -> int:
        return len(self.routes)

    @property
    def regions(self) -> list[list[tuple[int, int]]]:
        """``[[(asset_slot, draw_mode), ...], ...]`` indexed by region id."""
        return self.tables.scene_regions(self.scene, self.original)

    @property
    def draw_modes(self) -> dict[int, int]:
        return self.tables.scene_draw_modes(self.scene, self.original)

    # -- collision ---------------------------------------------------------

    def colisets(self) -> tuple | None:
        """``(common, per_scene)`` -- the two `coli/` files this scene loads.

        `ColiLoadForScene` (0x0048A3B0) loads `coli0.bin` for every scene plus
        `coli<scene+1>.bin`, and it guards `0 <= scene < 7`; scenes outside
        that range get no collision and this returns None.
        """
        if self._colisets is None:
            try:
                common_name, scene_name = colilib.scene_files(self.scene)
            except colilib.ColiError:
                self._colisets = ()
                return None
            d = self.game / "coli"
            try:
                self._colisets = (colilib.load(d / common_name),
                                  colilib.load(d / scene_name))
            except OSError:
                self._colisets = ()
        return self._colisets or None

    # -- event script ------------------------------------------------------

    def evt(self):
        """The parsed, relocated ``evt/`` table, or None if the scene has none."""
        if self._evt is None:
            name = self.evt_file
            if not name:
                return None
            path = self.game / "evt" / name
            if not path.exists():
                return None
            self._evt = evt.load(str(path), self.block_count)
        return self._evt

    # -- geometry ----------------------------------------------------------

    def geometry(self):
        """The authoritative geometry set for this stage.

        Globbing ``st<N>_*`` is wrong in both directions: it misses files the
        stage genuinely draws (``st3.bin``, and all of stage 6's reused
        ``st5_*`` geometry) and includes entries no region ever draws.

        The real set is the union of

          * every asset slot named by any of the scene's *regions* -- the
            sliding window the game streams and draws along the rail
            (:meth:`ExeTables.scene_regions`), and
          * every slot the event script loads with opcode ``0x50``.

        Whole-file loads (opcode ``0x52``) are deliberately excluded: those
        are spawnable actors -- enemies, characters -- instantiated at runtime
        from spawn descriptors, not placed scenery.

        With ``original`` set, the mode-1 region id tables are used instead --
        Original Mode. Region membership is byte-identical between the two
        modes; only a handful of id entries point at different models,
        swapping in the ``st_org00..st_org03`` files (and on stage 1,
        alternate ``st1_*`` entries) that Arcade Mode never draws. See
        docs/formats/pipeline.md.

        Returns ``(parts, model_regions, regions)`` where *parts* is the usual
        ``(name, models, bank)`` list, *model_regions* maps
        ``(part, model index)`` to the region ids that draw it, and *regions*
        is the raw region table.
        """
        if self._geometry is not None:
            return self._geometry

        game = self.game
        slots = self.tables.asset_slots()
        regions = self.regions

        wanted: dict[str, set[int]] = {}
        slot_regions: dict[tuple[str, int], set[int]] = {}
        for ri, region in enumerate(regions):
            for slot, _mode in region:
                rec = slots.get(slot)
                if not rec:
                    continue
                wanted.setdefault(rec[0], set()).add(rec[1])
                slot_regions.setdefault(rec, set()).add(ri)

        ev = self.evt()
        if ev is not None:
            for blk in ev.blocks:
                if blk.offset < 0:
                    continue
                for prog in blk.programs:
                    for ins in prog:
                        if ins.opcode in evt.SLOT_OPCODES and ins.raw:
                            rec = slots.get(ins.raw[0])
                            if rec:
                                wanted.setdefault(rec[0], set()).add(rec[1])

        draw_modes = self.draw_modes
        slot_of: dict[tuple[str, int], int] = {v: k for k, v in slots.items()}
        parts, model_regions = [], {}
        for fname in sorted(wanted):
            stem = fname[:-4] if fname.endswith(".bin") else fname
            pol = game / "pol" / fname
            if not pol.exists():
                continue
            cont = C.load(pol.read_bytes())
            bank = _bank_for(game, stem)
            models, order = [], sorted(wanted[fname])
            for entry in order:
                if entry >= cont.model_count:
                    continue
                try:
                    got = nl1.parse(cont.model(entry))
                except Exception:
                    continue
                for m in (got if isinstance(got, list) else [got]):
                    slot_id = slot_of.get((fname, entry))
                    model_regions[(stem, len(models))] = {
                        "regions": sorted(slot_regions.get((fname, entry), ())),
                        "draw_mode": (draw_modes.get(slot_id, 0)
                                      if slot_id is not None else 0),
                        "slot": slot_id,
                        "entry": entry,
                    }
                    models.append(m)
            if models:
                parts.append((stem, models, bank))
        self._geometry = (parts, model_regions, regions)
        return self._geometry

    # -- cameras -----------------------------------------------------------

    def cam_files(self):
        """``[cp_st<N>, op_st<N>]``, parsed."""
        if self._cam_files is None:
            self._cam_files = load_cam_paths(self.game, self.stage, None)
        return self._cam_files

    def campaths(self) -> CamPaths:
        """The stage's ``cam/`` paths keyed by their global slot id."""
        if self._campaths is None:
            self._campaths = CamPaths(self.tables, self.cam_files())
        return self._campaths

    # -- description -------------------------------------------------------

    def region_json(self) -> list[list[dict]]:
        """Region table with every slot resolved to a file and entry index."""
        slots = self.tables.asset_slots()
        return [[{"slot": s, "draw_mode": m,
                  "file": slots.get(s, ("?", 0))[0],
                  "entry": slots.get(s, ("?", 0))[1]} for s, m in reg]
                for reg in self.regions]

    def source_files(self) -> list[_Path]:
        """Every file on disk this stage's export is derived from.

        Recorded in the bundle manifest by SHA-256, so a bundle built from a
        different game build is detectable rather than mysteriously wrong.
        """
        out = [self.game / "Hod2.exe"]
        if self.evt_file:
            p = self.game / "evt" / self.evt_file
            if p.exists():
                out.append(p)
        com = self.game / "evt" / "comevtbl.bin"
        if com.exists():
            out.append(com)
        for cf in self.cam_files():
            p = self.game / "cam" / cf.name
            if p.exists():
                out.append(p)
        parts, _, _ = self.geometry()
        for stem, _models, _bank in parts:
            for sub in ("pol", "tex"):
                p = self.game / sub / f"{stem}.bin"
                if p.exists():
                    out.append(p)
        return out

File: web/src/hod2lib/stage.ts
/**
 * What a *stage* is: the single abstraction every exporter resolves against.
 * The port of `tools/hod2lib/stage.py`.
 *
 * A stage is not a directory and not a filename glob. It is a **scene id** --
 * the event system's own index -- and everything else follows from tables
 * compiled into `Hod2.exe`:
 *
 *     scene ---> evt/ file          0x00579928 -> 0x004D1C7C
 *           ---> route table        0x00597890      block flow graph
 *           ---> region tables      0x00576A2C/0x00576A5C  (mode 1: .../A8C)
 *           ---> geometry set       union of every region's asset slots,
 *                                   plus every slot opcode 0x50 loads
 *           ---> cam/ files         cp_st<N>, op_st<N>
 *
 * **Everything that touches a file is async here and synchronous in Python.**
 * That is the one structural difference in this package, and it is the seam:
 * a browser cannot block on a read. The laziness is the same -- constructing a
 * `Stage` reads nothing -- and each accessor caches its promise, so two callers
 * asking for the geometry at once decode it once.
 *
 * Reference: docs/formats/pipeline.md.
 */

import * as camlib from "./cam";
import { CamPaths } from "./campaths";
import * as colilib from "./coli";
import * as C from "./container";
import * as degraded from "./degraded";
import * as evt from "./evt";
import { ExeTables } from "./exetab";
import type { AssetSource } from "./io";
import * as nl1 from "./nl1";
import * as texbank from "./texbank";

/**
 * `g_GameMode` (0x009CA08C), as the EXE numbers it.
 *
 * The bundle used to emit a *different* number under the same name -- a flag,
 * `1 if original else 0` -- so a bundle's 0 meant the EXE's 2 and every reader
 * had to know which side of the seam it was on. There is one enumeration now;
 * `web/src/game/game_mode.ts` is its other half.
 *
 * **`ARCADE` was 2 until bundle format 7, and 2 is Training.** The values are
 * the title menu's own row order, which `TitleMenuRegisterSprites`
 * (`FUN_004962C0`) registers by name -- `tex\arcade00`, `tex\original_00`,
 * `tex\traning_00`, `tex\boss_00` -- and `TitleMenuUpdateAndSelect`
 * (`FUN_00496960`) writes the highlighted row straight into the global. Every
 * other writer in the program stores 0, so 1, 2 and 3 can only come from that
 * menu.
 */
export enum GameMode {
  /**
   * Arcade -- the coin-op game. Menu row 0, and the value every non-menu
   * writer of `g_GameMode` stores, so it is also the default.
   */
  ARCADE = 0,
  /** The story campaign. Menu row 1. */
  ORIGINAL = 1,
  /**
   * Training. Menu row 2. `ResetGameOnStart` sends it to scene 6, which is
   * `trnevtbl.bin`, and `g_training_lesson` (`0x009C9118`) selects the lesson.
   * No stage bundle is exported in this mode.
   */
  TRAINING = 2,
  /**
   * Boss. Menu row 3. `ResetGameOnStart` sends it to scene 0 block 0x10 and
   * `PreloadScreenAssetList` gives it a six-entry per-scene table of its own.
   * No shipped stage script is entered in this mode.
   */
  BOSS = 3,
}

/** stage number -> scene id, the event system's own index. */
export const STAGE_TO_SCENE: Record<number, number> =
  { 1: 0, 2: 1, 3: 2, 4: 3, 5: 4, 6: 5 };
export const SCENE_TO_STAGE: Record<number, number> =
  { 0: 1, 1: 2, 2: 3, 3: 4, 4: 5, 5: 6 };

/**
 * Parsed `Hod2.exe` tables per source, cached.
 *
 * The reference implementation caches on the game directory string because
 * that is its identity; here the {@link AssetSource} *is* the identity, and it
 * may be a directory handle with no path at all.
 */
const TABLES = new WeakMap<AssetSource, Promise<ExeTables | null>>();

export function getTables(source: AssetSource): Promise<ExeTables | null> {
  let p = TABLES.get(source);
  if (!p) {
    p = (async () => {
      if (!await source.exists("Hod2.exe")) return null;
      return ExeTables.create(await source.read("Hod2.exe"), "Hod2.exe");
    })();
    TABLES.set(source, p);
  }
  return p;
}

/** One `pol/` file's contribution to a stage: `[name, models, bank]`. */
export type Part = [string, nl1.Model[], texbank.Bank | null];

/**
 * The `tex/` bank paired with a `pol/` file, or null.
 *
 * The EXE's descriptor table is authoritative; the prefix-sum solver is only a
 * fallback for banks it does not list.
 */
async function bankFor(source: AssetSource, tables: ExeTables | null,
                       stem: string): Promise<texbank.Bank | null> {
  const tex = `tex/${stem}.bin`;
  if (!await source.exists(tex)) return null;
  const raw = await source.read(tex);
  const tc = C.load(raw);
  const data = tc.kind === C.COMPRESSED ? tc.data : raw;
  const entries = tables ? tables.entries(stem) : [];
  if (entries.length) return texbank.bankFromExe(data, entries);
  return null;
}

/** `[models, bank]` for a `pol/` asset and its `tex/` bank. */
export async function loadAsset(source: AssetSource, tables: ExeTables | null,
                                name: string):
    Promise<[nl1.Model[], texbank.Bank | null]> {
  const pol = `pol/${name}.bin`;
  if (!await source.exists(pol)) {
    throw new Error(`no such asset: ${pol}`);
  }

  const c = C.load(await source.read(pol));
  if (!c.models.length) throw new Error(`${name}: no models (${c.kind})`);
  const models = nl1.parseContainer(c);

  let bank: texbank.Bank | null = null;
  const tex = `tex/${name}.bin`;
  if (await source.exists(tex)) {
    const raw = await source.read(tex);
    const tc = C.load(raw);
    const data = tc.kind === C.COMPRESSED ? tc.data : raw;
    const entries = tables ? tables.entries(name) : [];
    if (entries.length) {
      bank = texbank.bankFromExe(data, entries);
    } else {
      const descs = texbank.harvestDescriptors(models);
      if (descs.size) bank = texbank.solveLayout(data, descs);
    }
  }
  return [models, bank];
}

/**
 * The `cam/` files belonging to a stage: `cp_stN` then `op_stN`.
 *
 * Returns `[]` when the asset has no matching camera file, which is normal --
 * only the six stages and a handful of cutscenes have one.
 */
export async function loadCamPaths(source: AssetSource, stage: number | null,
                                   name: string | null):
    Promise<camlib.CamFile[]> {
  let stems: string[] = [];
  if (stage !== null) {
    stems = [`cp_st${stage}`, `op_st${stage}`];
  } else if (name) {
    const m = /^st(\d+)_/.exec(name);           // st2_07 -> stage 2
    if (m) stems = [`cp_st${m[1]}`, `op_st${m[1]}`];
  }

  const out: camlib.CamFile[] = [];
  for (const stem of stems) {
    const p = `cam/${stem}.bin`;
    if (await source.exists(p)) {
      out.push(camlib.parse(await source.read(p), `${stem}.bin`));
    }
  }
  return out;
}

export interface ModelRegion {
  regions: number[];
  draw_mode: number;
  slot: number | null;
  entry: number;
}

export interface Geometry {
  parts: Part[];
  /** Keyed by {@link pairKey}: the stem and the model index. */
  modelRegions: Map<string, ModelRegion>;
  regions: [number, number][][];
}

/**
 * The key `modelRegions` and the slot inverse use for a `(name, n)` pair.
 *
 * The reference implementation keys those two maps on a Python tuple, which
 * JavaScript has no equivalent of, so the pair is spelled as one string with a
 * separator no part name can contain. Every reader of either map goes through
 * this function: `gltf` looked the same key up with a space in it once, and
 * the result was every model node in the bundle losing its region list -- a
 * bundle that loads, draws, and shows the whole stage at once.
 */
export function pairKey(a: string, b: number): string {
  return `${a}\u0000${b}`;
}

/**
 * One stage of the game, resolved from `Hod2.exe`.
 *
 * Construct with {@link Stage.create}: reading the EXE tables is the one
 * eager step, and it is asynchronous.
 */
export class Stage {
  readonly stage: number | null;
  readonly gameMode: number;

  private geometryP: Promise<Geometry> | null = null;
  private camFilesP: Promise<camlib.CamFile[]> | null = null;
  private campathsP: Promise<CamPaths> | null = null;
  private evtP: Promise<evt.EvtFile | null> | null = null;
  private colisetsP: Promise<[colilib.ColiFile, colilib.ColiFile] | null> | null
    = null;

  private constructor(
    readonly source: AssetSource,
    readonly scene: number,
    stage: number | null,
    readonly original: boolean,
    readonly tables: ExeTables,
  ) {
    this.stage = stage;
    this.gameMode = original ? GameMode.ORIGINAL : GameMode.ARCADE;
  }

  static async create(source: AssetSource,
                      opts: { stage?: number | null; scene?: number | null;
                              original?: boolean }): Promise<Stage> {
    let scene = opts.scene ?? null;
    const stageNo = opts.stage ?? null;
    if (scene === null) {
      if (stageNo === null) throw new Error("give stage or scene");
      scene = STAGE_TO_SCENE[stageNo] ?? null;
      if (scene === null) throw new Error(`stage ${stageNo} has no scene`);
    }
    const tables = await getTables(source);
    if (tables === null) {
      throw new Error(
        `Hod2.exe is required to resolve a stage: ${source.label}`);
    }
    return new Stage(source, scene,
                     stageNo !== null ? stageNo : (SCENE_TO_STAGE[scene] ?? null),
                     opts.original ?? false, tables);
  }

  // -- identity ---------------------------------------------------------

  /** Bundle/export name: `stage2`, or `stage2_original`. */
  get name(): string {
    const base = this.stage ? `stage${this.stage}` : `scene${this.scene}`;
    return base + (this.original ? "_original" : "");
  }

  get evtFile(): string | null {
    return this.tables.sceneEvtFile(this.scene);
  }

  /** `[[kind, next0, next1, next2], ...]` -- the block flow graph. */
  get routes(): [number, number, number, number][] {
    return this.tables.sceneRoutes(this.scene);
  }

  /**
   * `[terminal block, next scene's entry block]` for every ending.
   *
   * See {@link ExeTables.sceneExits} -- the terminal route record's `next[0]`
   * is where the next stage starts.
   */
  get exits(): [number, number][] {
    return this.tables.sceneExits(this.scene);
  }

  /**
   * Every block this scene can be entered at, ascending.
   *
   * See {@link ExeTables.sceneEntryBlocks}. Stage 3 and stage 4 have two;
   * every other stage has one.
   */
  get entries(): number[] {
    return this.tables.sceneEntryBlocks(this.scene);
  }

  get blockCount(): number {
    return this.routes.length;
  }

  /** `[[[assetSlot, drawMode], ...], ...]` indexed by region id. */
  get regions(): [number, number][][] {
    return this.tables.sceneRegions(this.scene, this.original);
  }

  get drawModes(): Map<number, number> {
    return this.tables.sceneDrawModes(this.scene, this.original);
  }

  // -- collision ---------------------------------------------------------

  /**
   * `[common, perScene]` -- the two `coli/` files this scene loads.
   *
   * `ColiLoadForScene` loads `coli0.bin` for every scene plus
   * `coli<scene+1>.bin`, and it guards `0 <= scene < 7`; scenes outside that
   * range get no collision and this returns null.
   */
  colisets(): Promise<[colilib.ColiFile, colilib.ColiFile] | null> {
    if (this.colisetsP === null) {
      this.colisetsP = (async () => {
        let names: [string, string];
        try {
          names = colilib.sceneFiles(this.scene);
        } catch (exc) {
          // not-a-loss: `ColiLoadForScene` guards `0 <= scene < 7` and this is
          // that guard. A scene outside the range has no collision *in the
          // engine*, so null is the faithful answer rather than a failure.
          if (!(exc instanceof colilib.ColiError)) throw exc;
          return null;
        }
        try {
          return [await colilib.load(this.source, `coli/${names[0]}`, names[0]),
                  await colilib.load(this.source, `coli/${names[1]}`, names[1])];
        } catch (exc) {
          // An install missing `coli/` exported a bundle with no collision in
          // it and said nothing: every wall in the stage becomes passable,
          // every ground query answers zero, and it reads as a gameplay bug
          // rather than a missing file.
          degraded.note("hod2lib.stage.colisets",
                        `coli/${names[0]} and coli/${names[1]}`,
                        "no collision for this scene at all", exc);
          return null;
        }
      })();
    }
    return this.colisetsP;
  }

  // -- event script ------------------------------------------------------

  /** The parsed, relocated `evt/` table, or null if the scene has none. */
  evt(): Promise<evt.EvtFile | null> {
    if (this.evtP === null) {
      this.evtP = (async () => {
        const name = this.evtFile;
        if (!name) return null;
        const path = `evt/${name}`;
        if (!await this.source.exists(path)) return null;
        // The shared 0x200-byte buffer that sits immediately below the stage
        // table in memory. `spawn_simple` points into it, so a stage table
        // that cannot see it resolves six of the game's seven screen-furniture
        // records to nothing. See `EvtFile.resolve`.
        const comPath = "evt/comevtbl.bin";
        const com = await this.source.exists(comPath)
          ? new evt.EvtFile(await this.source.read(comPath), "comevtbl.bin")
          : null;
        return evt.parse(await this.source.read(path), name, this.blockCount,
                         com);
      })();
    }
    return this.evtP;
  }

  // -- geometry ----------------------------------------------------------

  /**
   * The authoritative geometry set for this stage.
   *
   * Globbing `st<N>_*` is wrong in both directions: it misses files the stage
   * genuinely draws (`st3.bin`, and all of stage 6's reused `st5_*` geometry)
   * and includes entries no region ever draws.
   *
   * The real set is the union of every asset slot named by any of the scene's
   * *regions* -- the sliding window the game streams and draws along the rail
   * -- and every slot the event script loads with opcode `0x50`. Whole-file
   * loads (opcode `0x52`) are deliberately excluded: those are spawnable
   * actors instantiated at runtime from spawn descriptors, not placed scenery.
   */
  geometry(): Promise<Geometry> {
    if (this.geometryP === null) this.geometryP = this.buildGeometry();
    return this.geometryP;
  }

  private async buildGeometry(): Promise<Geometry> {
    const source = this.source;
    const slots = this.tables.assetSlots();
    const regions = this.regions;

    const wanted = new Map<string, Set<number>>();
    const slotRegions = new Map<string, Set<number>>();
    regions.forEach((region, ri) => {
      for (const [slot] of region) {
        const rec = slots.get(slot);
        if (!rec) continue;
        let s = wanted.get(rec[0]);
        if (!s) { s = new Set(); wanted.set(rec[0], s); }
        s.add(rec[1]);
        const key = pairKey(rec[0], rec[1]);
        let r = slotRegions.get(key);
        if (!r) { r = new Set(); slotRegions.set(key, r); }
        r.add(ri);
      }
    });

    const ev = await this.evt();
    if (ev !== null) {
      for (const blk of ev.blocks) {
        if (blk.offset < 0) continue;
        for (const prog of blk.programs) {
          for (const ins of prog) {
            if (evt.SLOT_OPCODES.includes(ins.opcode) && ins.raw.length) {
              const rec = slots.get(ins.raw[0]);
              if (rec) {
                let s = wanted.get(rec[0]);
                if (!s) { s = new Set(); wanted.set(rec[0], s); }
                s.add(rec[1]);
              }
            }
          }
        }
      }
    }

    const drawModes = this.drawModes;
    const slotOf = new Map<string, number>();
    for (const [k, v] of slots) slotOf.set(pairKey(v[0], v[1]), k);
    const parts: Part[] = [];
    const modelRegions = new Map<string, ModelRegion>();
    for (const fname of [...wanted.keys()].sort()) {
      const stem = fname.endsWith(".bin") ? fname.slice(0, -4) : fname;
      const pol = `pol/${fname}`;
      if (!await source.exists(pol)) continue;
      const cont = C.load(await source.read(pol));
      const bank = await bankFor(source, this.tables, stem);
      const models: nl1.Model[] = [];
      const order = [...wanted.get(fname)!].sort((a, b) => a - b);
      for (const entry of order) {
        if (entry >= cont.models.length) continue;
        let got: nl1.Model;
        try {
          got = nl1.parse(C.model(cont, entry));
        } catch (exc) {
          degraded.note("hod2lib.stage.geometry", `model ${entry} of ${fname}`,
                        "that model is absent from the geometry", exc);
          continue;
        }
        const slotId = slotOf.get(pairKey(fname, entry));
        modelRegions.set(pairKey(stem, models.length), {
          regions: [...(slotRegions.get(pairKey(fname, entry)) ?? [])]
            .sort((a, b) => a - b),
          draw_mode: slotId !== undefined ? (drawModes.get(slotId) ?? 0) : 0,
          slot: slotId ?? null,
          entry,
        });
        models.push(got);
      }
      if (models.length) parts.push([stem, models, bank]);
    }
    return { parts, modelRegions, regions };
  }

  // -- cameras -----------------------------------------------------------

  /** `[cp_st<N>, op_st<N>]`, parsed. */
  camFiles(): Promise<camlib.CamFile[]> {
    if (this.camFilesP === null) {
      this.camFilesP = loadCamPaths(this.source, this.stage, null);
    }
    return this.camFilesP;
  }

  /** The stage's `cam/` paths keyed by their global slot id. */
  campaths(): Promise<CamPaths> {
    if (this.campathsP === null) {
      this.campathsP = (async () =>
        new CamPaths(this.tables, await this.camFiles()))();
    }
    return this.campathsP;
  }

  // -- description -------------------------------------------------------

  /** Region table with every slot resolved to a file and entry index. */
  regionJson(): Record<string, unknown>[][] {
    const slots = this.tables.assetSlots();
    return this.regions.map((reg) => reg.map(([s, m]) => {
      const rec = slots.get(s) ?? ["?", 0] as [string, number];
      return { slot: s, draw_mode: m, file: rec[0], entry: rec[1] };
    }));
  }

  /**
   * Every file this stage's export is derived from, as source-relative paths.
   *
   * Recorded in the bundle manifest by SHA-256, so a bundle built from a
   * different game build is detectable rather than mysteriously wrong.
   */
  async sourceFiles(): Promise<string[]> {
    const out = ["Hod2.exe"];
    if (this.evtFile) {
      const p = `evt/${this.evtFile}`;
      if (await this.source.exists(p)) out.push(p);
    }
    const com = "evt/comevtbl.bin";
    if (await this.source.exists(com)) out.push(com);
    for (const cf of await this.camFiles()) {
      const p = `cam/${cf.name}`;
      if (await this.source.exists(p)) out.push(p);
    }
    const { parts } = await this.geometry();
    for (const [stem] of parts) {
      for (const sub of ["pol", "tex"]) {
        const p = `${sub}/${stem}.bin`;
        if (await this.source.exists(p)) out.push(p);
      }
    }
    return out;
  }
}
